File: 143-reorder-list/reorder-list.c
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     struct ListNode *next;
 * };
 */
  struct ListNode*reverse(struct ListNode*head){
   
    struct ListNode*prev=NULL;
    struct ListNode*curr=head;
    while(curr){
        struct ListNode*next=curr->next;
        curr->next=prev;
        prev=curr;
        curr=next;
    }
    return prev;
  }
void reorderList(struct ListNode* head) {
   if(!head||!head->next||!head->next->next) return;
   struct ListNode* slow=head;
   struct ListNode* fast=head;
   while(fast&&fast->next&&fast->next->next){
    slow=slow->next;
    fast=fast->next->next;
   } 
   struct ListNode*second=reverse(slow->next);
   slow->next=NULL;
   struct ListNode*first=head;
   while(second){
    struct ListNode*temp1=first->next;
    struct ListNode*temp2=second->next;
    first->next=second;
    second->next=temp1;
    first=temp1;
    second=temp2;
   }
}

File: 143-reorder-list/reorder-list.py
# Definition for singly-linked list.
# class ListNode(object):
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next
class Solution(object):
    def reorderList(self, head):
        if not head or not head.next:
            return head
        slow=fast= head
        while fast and fast.next:
            slow=slow.next
            
            fast=fast.next.next
        prev,curr=None,slow.next
        slow.next=None
        while curr:
            nxt=curr.next
            curr.next=prev
            prev=curr
            curr=nxt
         
        
        first,second=head,prev
        while second:
            tmp1=first.next
            tmp2=second.next
            first.next=second
            second.next=tmp1
            first,second=tmp1,tmp2

            

         
        